libs/constrained-generators/src/Constrained/Generation.hs

```diff
@@ -0,0 +1,1208 @@
+        newStages = concatMap (newStage spec) ps
+        plan' = foldr mergeSolverStage plan newStages
+        -- Note use of the Term Pattern Equal
+        newStage specl (Assert (Equal (V x') t)) =
+          termVarEqCases specl x' t
+        newStage specr (Assert (Equal t (V x'))) =
+          termVarEqCases specr x' t
+        newStage _ _ = []
+
+        termVarEqCases :: HasSpec b => Specification a -> Var b -> Term b -> [SolverStage]
+        termVarEqCases (MemberSpec vs) x' t
+          | Set.singleton (Name x) == freeVarSet t =
+              [SolverStage x' [] $ MemberSpec (NE.nub (fmap (\v -> errorGE $ runTerm (singletonEnv x v) t) vs))]
+        termVarEqCases specx x' t
+          | Just Refl <- eqVar x x'
+          , [Name y] <- Set.toList $ freeVarSet t
+          , Result ctx <- toCtx y t =
+              [SolverStage y [] (propagateSpec specx ctx)]
+        termVarEqCases _ _ _ = []
+
+data EqW :: [Type] -> Type -> Type where
+  EqualW :: (Eq a, HasSpec a) => EqW '[a, a] Bool
+
+deriving instance Eq (EqW dom rng)
+
+instance Show (EqW d r) where
+  show EqualW = "==."
+
+instance Syntax EqW where
+  isInfix EqualW = True
+
+instance Semantics EqW where
+  semantics EqualW = (==)
+
+instance Logic EqW where
+  propagate f ctxt (ExplainSpec es s) = explainSpec es $ propagate f ctxt s
+  propagate _ _ TrueSpec = TrueSpec
+  propagate _ _ (ErrorSpec msgs) = ErrorSpec msgs
+  propagate EqualW (HOLE :? Value x :> Nil) (SuspendedSpec v ps) =
+    constrained $ \v' -> Let (App EqualW (v' :> Lit x :> Nil)) (v :-> ps)
+  propagate EqualW (Value x :! Unary HOLE) (SuspendedSpec v ps) =
+    constrained $ \v' -> Let (App EqualW (Lit x :> v' :> Nil)) (v :-> ps)
+  propagate EqualW (HOLE :? Value s :> Nil) spec =
+    caseBoolSpec spec $ \case
+      True -> equalSpec s
+      False -> notEqualSpec s
+  propagate EqualW (Value s :! Unary HOLE) spec =
+    caseBoolSpec spec $ \case
+      True -> equalSpec s
+      False -> notEqualSpec s
+
+  rewriteRules EqualW (t :> t' :> Nil) Evidence
+    | t == t' = Just $ lit True
+    | otherwise = Nothing
+
+  -- TODO: fixme
+  saturate EqualW (FromGeneric (InjLeft _) :> t :> Nil) = [toPreds t (SumSpec Nothing TrueSpec (ErrorSpec (pure "saturatePred")))]
+  saturate EqualW (FromGeneric (InjRight _) :> t :> Nil) = [toPreds t (SumSpec Nothing (ErrorSpec (pure "saturatePred")) TrueSpec)]
+  saturate _ _ = []
+
+infix 4 ==.
+
+(==.) :: HasSpec a => Term a -> Term a -> Term Bool
+(==.) = appTerm EqualW
+
+pattern Equal ::
+  forall b.
+  () =>
+  forall a.
+  (b ~ Bool, Eq a, HasSpec a) =>
+  Term a ->
+  Term a ->
+  Term b
+pattern Equal x y <-
+  ( App
+      (getWitness -> Just EqualW)
+      (x :> y :> Nil)
+    )
+
+whenTrue :: forall p. IsPred p => Term Bool -> p -> Pred
+whenTrue (Lit True) (toPred -> p) = p
+whenTrue (Lit False) _ = TruePred
+-- TODO: fixme
+whenTrue b (toPred -> FalsePred {}) = assert (not_ b)
+whenTrue _ (toPred -> TruePred) = TruePred
+whenTrue b (toPred -> p) = When b p
```

Comment:
What needs fixing here? Is this something you wrote, or is it left over from some previous PR?